Balance Law: add docs, refactor some of the interface, add AtmosDycore tests

jkozdon opened the pull request.

simonbyrne commented on src/Atmos/Dycore/test/isentropic_vortex.jl:
```diff
@@ -19,6 +19,8 @@ using CLIMA.PlanetParameters: R_d, cp_d, grav, cv_d
 
 const halfperiod = 5
 
+const print_diagnostics = length(ARGS) == 0 || parse(Bool, ARGS[1])
```
we should use Julia's logging infrastructure for this `@debug`, `@info` for this: want me to have a go?

jkozdon replied:
Sure. I was just trying to suppress the existing output in the tests

jkozdon replied:
We need it to play well with MPI too (so best would be also some `@root_info` as well in which case only 1 ranks does the output not everyone)

jkozdon replied:
You can see how I used this down on line 127 to define an `io` variable.

simonbyrne replied:
I think we should just be able to set a different log level on each rank.

jkozdon replied:
really this is related to larger issue #134

simonbyrne replied:
theoretically you could set up an MPILogger or something like that which would handle everything automatically, but this will work for now.

jkozdon replied:
Sounds good. Are you adding it to this pull request or will you do a separate one?

simonbyrne replied:
I can just add it to this one.

jkozdon replied:
ok

simonbyrne replied:
Let me know what you think of ce11cbdd510f368cc6733f465c3d2d8e39226917: we could move some of this into Utilities proper.

simonbyrne replied:
Okay, I've updated all of them, hopefully that should work nicely.

jkozdon commented:
Forced a push to update to current master and try to fix the windows test

jkozdon commented:
Overall I'm good with this for now (assuming the tests pass)

jkozdon commented:
My fix didn't work. Reverting.

The problem is that `-env` is not a standard option for `mpiexec` (my mac which uses openmpi 3.0.0 doesn't recognize it either).

simonbyrne commented:
Since the tests are all run locally, it might suffice to set the ENV from
within Julia before calling MPI?

jkozdon commented:
Done. I just set this for all the tests in `test/runtest.jl` as this seemed most reasonable to me (?). Uniform behaviour across all the tests with any logging output.

jkozdon commented:
I added the test splitting idea from #169 into this pull request. I know that you weren't a huge fan of it @simonbyrne so I can remove it if you want (only used right now on travis).

I think the test set idea might be useful in the future, since the tests I'm running right now are pretty short compared to what will be coming once we test the full system. I think we'll likely have a short and long test set. This allows folks to also split tests into integration tests, unit test, moist test, etc.

I also added another environment variable "JULIA_CLIMA_TESTING" to run separate tests when in testing versus running tests as stand alone. Doing this allows me to run one fewer resolution in CI.

simonbyrne commented:
@jkozdon I have a small fix: should I just push it to this branch?

jkozdon commented:
> @jkozdon I have a small fix: should I just push it to this branch?

Will merge commit from #181 shortly.

jkozdon commented:
@simonbyrne Just an FYI, we still probably want some sort of MPI-aware logging. Subtle bug that took me a bit to take down was an MPI call inside an `@sinfo` which of course only the root called. Really we want all the ranks to execute everything in the logging calls but only the root rank to do the print.

jkozdon commented:
Adding logging to #172

jkozdon commented:
@simonbyrne or @lcw anything else you want addressed here before this is merged?

jkozdon commented:
> @simonbyrne or @lcw anything else you want addressed here before this is merged?

I know it's turned into a bit of a beast of a PR, hopefully future ones will be more bitesized!

lcw commented:
I love the docs!  Looks good to me.

simonbyrne commented:
bors r+
